### PR TITLE
Add foldEvery

### PR DESCRIPTION
**Pull request by TimWSpence**
- Windows the stream into fixed widths and runs a fold incrementally
  on each window, emitting the result of the fold

### CODE REVIEW

**Comment by TimWSpence** [comment]
Is there any interest in me cleaning this up? A colleague challenged me to write something like this recently and I thought it seemed a reasoanably useful and generic function

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 1]
```diff
@@ -3024,6 +3024,39 @@ object Stream extends StreamLowPriority {
         .eval(timer.clock.monotonic(NANOSECONDS).map(now => (now - start).nanos))
     }
 
+  def foldEvery[F[_], A, B](
+      d: FiniteDuration
+  )(init: B)(f: (B, A) => B)(implicit F: Sync[F]): Pipe[F, A, B] = { in =>
+    {
+      val width = d.toMillis
+
+      val signal = Stream.eval(F.delay(System.currentTimeMillis())).repeat.zipWithPrevious.map {
```
Comment (Daenyth) [inline core/shared/src/main/scala/fs2/Stream.scala]: This should use `Timer[F]` rather than `System`

Comment (TimWSpence) [reply core/shared/src/main/scala/fs2/Stream.scala]: For sure, that would definitely be part of cleanup! I just thought I'd get some feedback on whether people would like it to be part of fs2, before I invest time in cleaning it up, writing tests, etc

**Comment by mpilquist** [comment]
Is `s.foldEvery(d)(z)(f) == s.groupWithin(Int.MaxValue, d).map(_.fold(z)(f))`?

**Comment by CremboC** [comment]
@mpilquist in theory yes, but the key difference is the stream nature.
`groupWithin` will basically block for either `n` elements, or `d` time and _then_ do the aggregation afterwards (basically wasting time, blocking the upstream then)
while `foldEvery` will aggregate in a streaming fashion, without blocking upstream. Does this make sense?

We also implemented `evalFoldEvery` with @TimWSpence which I think is worth adding to this PR.

**Pull request closed without merging**